pkgs/native_assets_builder/doc/protocol/v1.md

```diff
@@ -0,0 +1,216 @@
+
+# Target operating system
+#
+# Combined with `target_architecture` this specifies the "target" for which
+# assets should be built.
+target_os: android | ios | linux | macos | windows
+
+# Version of this file.
+version: 1.0.0
+```
+
+### `build_output.yaml`
+Mapping from `assetId`s to files created by `build.dart`.
+
+```yaml
+# The list of assets.
+#
+# In dry runs, must contain assets for each architecture for the requested os.
+assets:
+  - id: 'package:my_package_with_native_assets/src/foo.dart'
+    link_mode: dynamic | static | prefer-dynamic | prefer-static
+    path:
+      path_type: absolute | system | process | executable
+      # Only provided for path_type absolute and system.
+      #
+      # If path_type absolute: The absolute path to the file name.
+      #
+      # If path_type system: The path of the dynamic library as available on 
+      # the target machine's PATH.
+      uri: /absolute/path/to/outdir/arbitrary_filename.whatever
+    target: linux_x64
+  ...
+
+# The files used by this build.
+#
+# If any of the files in [dependencies2] are modified after [timestamp], the
+# build will be re-run.
+#
+# Not output on dry runs.
+dependencies:
+  - /absolute/path/to/my_package_with_native_assets/build.dart
+  ...
+
+# The time the build this output is for started.
+#
+# Must be before any of the files in dependencies are read.
+timestamp: 2024-01-02 17:05:35.000
+
+# Metadata usable for build.dart of packages directly depending on this package.
+#
+# Not output in dry runs.
+metadata:
+  # Key value pairs.
+  some_key: some_value
+
+# Version of this file.
+version: 1.0.0
+```
+
+
+### `.dart_tool/native_assets.yaml`
+
+This file is internal to the Dart SDK and Flutter SDK.
+It produced by `native_asset_builder` (as embedded in the SDKs).
+For each `target` it maps from `assetId` to `url`
+(with `urlKind` indicating if it's a relative or absolute path).
+
+```yaml
+# Version of this file.
+format-version: [1, 0, 0]
+
+# A mapping from target and assetId to an asset.
+native-assets:
+  # A <target>.
+  linux_x64:
+    # An <assetId>.
+    : 'package:my_package_with_native_assets/src/foo.dart'
+      # A list of path_type and optionally a path.
+      - absolute | relative | system | process | executable
+      - <url> # Only provided for absolute, relative, and system path types.
+  ...
+```
+
+This file is used by the embedded SDK to resolve `assetId`s when running in
+development mode or build a deployable application (or standalone executable).
+
+## Wishes to improve for v2
+
+* Multiple asset types. Rename `Asset` to `NativeCodeAsset` and introduce `DataAsset`s.
+* The path_types really only make sense for `link_mode: dynamic`.
+* Fix the mismatch between `target` for output but `os` and `architecture` for config.
+  * Should the architecture field of native code assets be optional and be ommitted in dry runs?
+* Introduce a `link.dart` protocol with a link input and link output.
```

Comment:
I would rather have a v1 of that as well. To be discussed elsewhere though.

Reply:
Because what we do with `link.dart` influences `build.dart`, I think we need a holistic design doc. (Does the build.dart output an `AssetForLink`, or `List<Asset> assetsForLinking`, do we introduce unlimited intermediary asset types for linking, or that can subtype data assets. etc.)